tokio-util/src/sync/cancellation_token.rs

```diff
@@ -222,3 +258,72 @@ impl<'a> Future for WaitForCancellationFuture<'a> {
         }
     }
 }
+
+// ===== impl WaitForCancellationFutureOwned =====
+
+impl core::fmt::Debug for WaitForCancellationFutureOwned {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("WaitForCancellationFutureOwned").finish()
+    }
+}
+
+impl WaitForCancellationFutureOwned {
+    fn new(cancellation_token: CancellationToken) -> Self {
+        let future = cancellation_token.inner.notified();
+
+        // Safety:
+        //  - We transmute notified into Notified<'static>
+        //    since there's no 'self
+        //    Since future is declared before cancellation_token,
+        //    it will be dropped before cancellation_token so it
+        //    will always holds a valid reference.
+        //  - this.cancellation_token contains an Arc to TreeNode,
+        //    which is guaranteed to have stable dereference.
+        //    So it's ok to move it here as long as `Notified::enabled`
+        //    or `Notified::poll` is not called.
+        let future: tokio::sync::futures::Notified<'static> = unsafe { mem::transmute(future) };
```

Comment:
You've duplicated this. Can you put it into a helper method?

Additionally, I have a slight preference for obtaining the future like this:
```Rust
let inner_ptr = Arc::as_ptr(&cancellation_token.inner);
let future = (*inner_ptr).notified();
```
Why? Because the documentation for `Arc::as_ptr` says the following:

> The pointer is valid for as long as there are strong counts in the Arc.

So there is no ambiguity about whether our references remains valid if the Arc is moved or anything like that. Additionally, this would let you avoid the transmute.

Reply:
Thanks!

I've extracted new unsafe fn `new_future` which uses `Arc::as_ptr`.